### ISSUE TITLE
Should /containers/(id)/update take HostConfig or only Resources

### ISSUE THREAD

**Issue by vdemeester** [docker/docker#18957]
The current implementation of `update`(#15078), and thus the endpoint `/containers/(id)/update`, is using `HostConfig` struct although it's only updating `Resources`. 

This raises few concerns :
- We decided that `update` should only *update* stuff that are in `Resources` struct, but it's possible to send a complete `HostConfig` struct which could lead to confusion — like why is some fields not taken care of, etc….
- If a complete `HostConfig` struct is passed, it will completely and **silently ignore** all attributes except `Resources`. It should at least put a warning.

I think we should *reduce* what we send and change the API `/containers/(id)/update` to just take `Resoures` struct instead of `HostConfig` — it would make it clear that only stuff *in* this struct are supported (and it's simpler to implement than doing some diff or something and warn that something that is not in `Resources` has been wrongly set).

/cc @hqhq @tiborvass @thaJeztah @runcom @duglin @unclejack @icecrime @calavera

**Comment by thaJeztah** [comment]
I think that, although we currently only accept "Resources" to be updated, that other options could be updatable as well in the future, so I'm -1 on passing a different struct than (e.g.) "create" accepts.

(I briefly discussed this with @vdemeester, and it's not sure if this is something easily implemented, and maintainable, but I'll post it here for the discussion):

(Ideally), anything that's passed in the struct that is *not* part of `Resources` (or *not* whitelisted as updatable) and is *different* from the current RunConfig, should probably result in an error.

Basically calculate a "diff" between what's passed and the current `HostConfig`, then check for each *modified* property if it's in the whitelist of mutable properties, and if not, produce an error.

**Comment by duglin** [comment]
I would prefer to look at this from a UX perspective. As a user of an API if there are properties in a struct that no one can ever set w/o them being ignored or generating an error then they shouldn't be there.  Each arg/struct for an API should be exactly what that API needs, not what some other API needs (and we're trying to reuse it) or what some internal structure/API needs.  I thought this was part of the reason we were headed down the path of creating all of the API structs in the first place - to separate what's sent over the wire from what we use internally, so each can change independently. 

Having said that, if two APIs share similar data then it would be good to make them look alike. So, if update() is a subset of create() then I think it make sense to have its struct be a subset (and look like a subset) of create's struct.  Then, as @thaJeztah mentions, if update's struct grows over time to look more like create's then we have a look-n-feel consistency that will be helpful to users.

So, if update() just wants what's in `Resource` (**and everything in `Resources` is valid**) and nothing from `HostConfig` then I'm +1 on this idea.

**Comment by hqhq** [comment]
I agree with @thaJeztah , it's highly likely that we'll support other fields out of `Resources`, and backward compatibility for this would be painful, we better not do this util we are 100% sure update will be limited to `Resources` only. And I would definitely vote for **NOT** limit this to `Resources` only :pray:

**Comment by aanand** [comment]
> (Ideally), anything that's passed in the struct that is not part of `Resources` (or not whitelisted as updatable) and is different from the current RunConfig, should probably result in an error.

From a backwards-compatibility standpoint, it might be better to produce a *warning*, rather than an error, but it depends on what sort of behaviour you want for newer scripts on older versions of Engine. If a program/script assumes that e.g. a container's `RestartPolicy` can be updated, should that program break if it actually can't be updated, or is it OK to continue?

If you *do* produce an error, it should definitely be made possible to detect that particular class of error on the client-side without resorting to string matching.

**Comment by thaJeztah** [comment]
@aanand fyi, the `docker update` feature is not yet in a release, and will be in 1.10, so there's no real backwards-compatibility issue yet. We would like to have this decision made before the release though :smile:

**Comment by aanand** [comment]
@thaJeztah yeah, I'm thinking more about future concerns as properties are added to the whitelist. Perhaps "forwards-compatibility" is the right term here: forwards-compatibility of Engine 1.10 with scripts written at a later date.

**Comment by vdemeester** [comment]
@aanand my idea is that to produce an error or a warning would be the last resort (meaning keeping `HostConfig`). I'm way more enclined to not use `HostConfig` struct for this one (see the related PR #19104).
